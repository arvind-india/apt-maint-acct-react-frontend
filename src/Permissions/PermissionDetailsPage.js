import React from 'react'
import { Redirect } from 'react-router-dom'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import Select from 'react-select';

import {
    Form,
    Button,
    FormGroup,
    FormText,
    Input,
    Label,
    Col,
    FormFeedback
} from 'reactstrap'

import { permissionActions as actions, alertActions } from '../_actions'
import { MODULES } from '../_constants'

class PermissionDetailsPage extends React.Component {

  constructor(props) {
    super(props)
console.log('props in constructor: ', props)
    const { dispatch, permissionDetails, match, location } = props
    let iModel = location.state.model // initial model
    let operations = iModel.operations?iModel.operations:''
    this.state = {
      ...iModel,
      submitted: false,
      touched: false,
      adding: match.params.id === "0",
      cPerm: operations.includes('C'), // create permission
      rPerm: operations.includes('R'), // read permission
      uPerm: operations.includes('U'), // update permission
      dPerm: operations.includes('D')  // delete permission
    }

    this.handleChange = this.handleChange.bind(this)
    this.handleOperationsChange = this.handleOperationsChange.bind(this)
//    this.handleResourceChange = this.handleResourceChange.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)

    // dispatch(actions.getAll())
    dispatch(alertActions.clear())  // clear alert messages from other pages
  }
  componentDidMount() {
    this.props.dispatch(actions.getById(this.props.match.params.id))
    console.log('props on componentDidMount: ', this.props)
  }
  handleSubmit(event) {
    event.preventDefault()
    this.setState({ submitted: true })

    const { mModel, adding } = this.state
    const { dispatch, permissionDetails } = this.props
    let modelDB = permissionDetails.data
    let canSave = this.canSave()
    let cProps = this.changedProps()
    if(adding) {
      modelDB.id = 0
    }
    if(canSave) {
      mModel.id = modelDB.id
    }
    if ( cProps.length == 0 ) {
      dispatch(alertActions.error('No changes found...'))
    } else if(canSave){
      dispatch(actions.saveChanges(mModel))
    } else {
      dispatch(alertActions.error('Missing data'))
    }

  }
  canSave() { // check for changes in mModel, if changes present, it can save
    const { mModel } = this.state
    if(mModel.resource == null){ // no changes in resource, then remove this attribute
      delete mModel.resource
    }
    for(const prop in mModel) {
      if(prop == 'resource' && mModel.resource == '')
        return true; // empty resource is valid
      if( !mModel[prop] ) {
        return false
      }
    }
    return true
  }
  changedProps() {
    const { mModel } = this.state
    const { permissionDetails } = this.props
    let modelDB = permissionDetails.data
    let props = []
    // check for changes in mModel props
    for(const prop in mModel) {
      if(modelDB[prop] != mModel[prop]) { // if data is changed wrt data in database
        props.push(prop)
      } else {
        delete mModel[prop]  // remove unchanged property
      }

    }
    return props;
  }
  handleChange(event) {
    const { name, value } = event.target
    this.setState({
        [name]: value
    })
  }
  handleOperationsChange(event) {
    const { name, value } = event.target
    const { cPerm, rPerm, uPerm, dPerm } = this.state
    // toggle state value
    switch(value) {
      case 'C': this.setState({ cPerm: !cPerm }); break
      case 'R': this.setState({ rPerm: !rPerm }); break
      case 'U': this.setState({ uPerm: !uPerm }); break
      case 'D': this.setState({ dPerm: !dPerm }); break
    }
  }
/*  handleResourceChange(resource) {
    // const { mModel } = this.state
console.log('resource: ', resource)
    this.setState({
      resource: resource
    })
  } */
  render() {
    const { permissionDetails, user, match, alert, submitted } = this.props
    let model = permissionDetails
    return (
      <div>
        <h2>Permission Details</h2>
        {alert.message && <div className={`alert ${alert.type}`}>{alert.message}</div>}
        {model.loading && <em>Loading model details...}</em>}
        {model.error && <span className="text-danger">{model.error}</span>}
        {model.data && this.show(model.data)}
      </div>
    )
  }
  show(data){
    let title = this.state.adding?'Add':'View or Edit'
    return <Form onSubmit={this.handleSubmit} className="grid-form">
      <fieldset>
  			<legend>{title}</legend>
        <div data-row-span="2">
          {this.showOperations(data)}
          {this.showResource(data)}
        </div>
        <div data-row-span="1">
          {this.showCondition(data)}
        </div>
        <div data-row-span="1">
          {this.showDescription(data)}
        </div>
      </fieldset>
      <br/>
      <Button type="submit" color="primary">Save</Button>
      <Button color="link"><Link to="/permissions">Cancel</Link></Button>
    </Form>
  }
  showOperations(data) {
    const { submitted, cPerm, rPerm, uPerm, dPerm } = this.state

    return <div data-field-span="1">
        <Label>Operations</Label>
        <FormGroup check inline>
          <Label check>
            <Input
              type="checkbox"
              name="operations"
              value="C"
              checked={cPerm}
              onChange={this.handleOperationsChange}
            /> Create
          </Label>
        </FormGroup>
        <FormGroup check inline>
          <Label check>
            <Input
              type="checkbox"
              name="operations"
              value="R"
              checked={rPerm}
              onChange={this.handleOperationsChange}
            /> Read
          </Label>
        </FormGroup>
        <FormGroup check inline>
          <Label check>
            <Input
              type="checkbox"
              name="operations"
              value="U"
              checked={uPerm}
              onChange={this.handleOperationsChange}
            /> Update
          </Label>
        </FormGroup>
        <FormGroup check inline>
          <Label check>
            <Input
              type="checkbox"
              name="operations"
              value="D"
              checked={dPerm}
              onChange={this.handleOperationsChange}
            /> Delete
          </Label>
        </FormGroup>
        {submitted && (cPerm || rPerm || uPerm || dPerm)
          && <FormText color="danger">Permission operations is required</FormText>}
      </div>
  }
  showResource(data) {
    const { submitted, resource } = this.state
    return <div data-field-span="1">
      <Label>Resource</Label>
      <Select
        name="form-field-name"
        value={resource}
        simpleValue={true}
        placeholder="Select Inherits..."
        onChange={this.handleChange}
        valueKey="name"
        labelKey="label"
        options={MODULES}
      />
      {submitted && resource != null && resource == ""
        && <FormText color="danger">Permission resource is required</FormText>}
     </div>
  }
  showCondition(data) {
    const { condition } = this.state
    return <div data-field-span="1">
				<Label>Conditions</Label>
        <Input
          type="text"
          name="condition"
          value={condition}
          placeholder="<conditional script here>"
          title="Short condition on Permission"
          className="inputField"
          onChange={this.handleChange}
        />
			</div>
  }
  showDescription(data) {
    const { submitted, description } = this.state
    return <div data-field-span="1">
				<Label>Description</Label>
        <Input
          type="text"
          name="description"
          value={description}
          placeholder="Description here"
          className="inputField"
          onChange={this.handleChange}
        />
        {submitted && description != null && description == ""
          && <FormText color="danger">Permission description is required</FormText>}
			</div>
  }
}

function mapStateToProps(state) {
  const { permissionDetails, authentication, alert } = state
  const { user } = authentication
  return {
    user,
    permissionDetails,
    alert
  }
}

const connectedDetailsPage = connect(mapStateToProps)(PermissionDetailsPage)
export { connectedDetailsPage as PermissionDetailsPage }
