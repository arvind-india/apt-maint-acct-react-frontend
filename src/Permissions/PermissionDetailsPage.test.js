// PermissionsPage.test.js

import React from 'react'
import { shallow } from 'enzyme'

import { PermissionDetails } from './PermissionDetailsPage'

const mockPermissions = [
  {
    id: 1,
    operations: 'R',
    resource: 'ACCOUNT',
    condition: '',
    description:'Grands Read-only permission on ACCOUNT',
  },
  {
    id: 2,
    operations: 'RU',
    resource: 'USER',
    condition: '',
    description:'Grands Read and Edit permissions on USER',
  }
]

const mockProps = [
  {
    alert: {message: "testingAlert"},
    authzn: {
      allowsAdd: true,
      allowsEdit: true,
      allowsDelete: true,
      condition: ''
    },
    clearAlert: jest.fn(),
    saveChanges: jest.fn(),
    error: jest.fn(),
    permissions: {
      items: []
    }
  },
  {
    alert: {message: "testingAlert"},
    authzn: {
      allowsAdd: true,
      allowsEdit: true,
      allowsDelete: true,
      condition: ''
    },
    clearAlert: jest.fn(),
    saveChanges: jest.fn(),
    error: jest.fn(),
    permissions: {
      items: []
    }
  }
]
const mockLocations = [
  {
    state: {
      model: mockPermissions[0],
      submitted: false,
      adding: false
    }
  },
  {
    state: {
      model: mockPermissions[1],
      submitted: false,
      adding: false
    }
  }
]

const mocks = [
  {
    operations: {name: 'operations', value: mockPermissions[0].operations},
    resource: mockPermissions[0].resource,
    condition: {name: 'condition', value: mockPermissions[0].condition},
    description: {name: 'description', value: mockPermissions[0].description}
  },
  {
    operations: {name: 'operations', value: mockPermissions[1].operations},
    resource: mockPermissions[1].resource,
    condition: {name: 'condition', value: mockPermissions[1].condition},
    description: {name: 'description', value: mockPermissions[1].description}
  }
]
const component = shallow(<PermissionDetails {...mockProps[0]} location={mockLocations[0]}/>)

describe('PermissionsPage params test', () => {
  it('should render PermissionsPage without crashing', () => {
    component
  })
  it('should render initial layout', () => {
    expect(component.getElements()).toMatchSnapshot()
  })
  it('should have model state', () => {
    expect(component.state('model')).toBeDefined()
  })
  it('should have submitted state', () => {
    expect(component.state('submitted')).toBeDefined()
  })
  it('should have adding state', () => {
    expect(component.state('adding')).toBeDefined()
  })
})

describe('PermissionDetailsPage events test 1', () => {
  it('should call clearAlert on submit and no crash', () => {
    component.find('#permissionDetailsForm').simulate('submit', {preventDefault() {}})
    expect(mockProps[0].clearAlert.mock.calls.length).toEqual(1)
  })
  it('should be called with required states as arguments', () => {
    component.find('.operations [checked=true]').simulate('change', {target: mocks[0].operations}) // toggles from 'R' to ''
    component.find('#readOperations').simulate('change', {target: mocks[0].operations}) // calling again to toggle from '' to 'R'
    component.find('Select').simulate('change', mocks[0].resource)
    component.find('#condition').simulate('change', {target: mocks[0].condition})
    component.find('#description').simulate('change', {target: mocks[0].description})
    component.find('#permissionDetailsForm').simulate('submit', {preventDefault() {}})
    expect(mockProps[0].saveChanges.mock.calls[1][0]).toEqual(mockPermissions[0])
  })

})

const component2 = shallow(<PermissionDetails {...mockProps[1]} location={mockLocations[1]}/>)
describe('PermissionDetailsPage events test 2', () => {
  it('should be called with required states as arguments', () => {
    // yet to explore below statement; it works!!! Even though 'Create' checkbox is searched and is simulated to 'RU' 
    component2.find('#createOperations').simulate('change', {target: mocks[1].operations})
    //component2.find('#readOperations').simulate('change', {target: mocks[1].operations})
    component2.find('Select').simulate('change', mocks[1].resource)
    component2.find('#condition').simulate('change', {target: mocks[1].condition})
    component2.find('#description').simulate('change', {target: mocks[1].description})
    component2.find('#permissionDetailsForm').simulate('submit', {preventDefault() {}})
    expect(mockProps[1].saveChanges.mock.calls[0][0]).toEqual(mockPermissions[1])
  })
})
