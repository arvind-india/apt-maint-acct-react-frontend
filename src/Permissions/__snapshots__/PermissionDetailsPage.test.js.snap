// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PermissionsPage params test should render initial layout 1`] = `
Array [
  <div>
    <h2>
        Permission Details
    </h2>
    <div
        className="alert undefined"
    >
        testingAlert
    </div>
    <Form
        className="grid-form"
        id="permissionDetailsForm"
        onSubmit={[Function]}
        tag="form"
    >
        <fieldset>
            <legend>
                Edit
            </legend>
            <div
                data-row-span="2"
            >
                <div
                    data-field-span="1"
                >
                    <Label
                        tag="label"
                        widths={
                            Array [
                                "xs",
                                "sm",
                                "md",
                                "lg",
                                "xl",
                              ]
                        }
                    >
                        Operations
                    </Label>
                    <FormGroup
                        check={true}
                        inline={true}
                        tag="div"
                    >
                        <Label
                            check={true}
                            tag="label"
                            widths={
                                Array [
                                    "xs",
                                    "sm",
                                    "md",
                                    "lg",
                                    "xl",
                                  ]
                            }
                        >
                            <Input
                                checked={false}
                                className="operations"
                                id="createOperations"
                                name="operations"
                                onChange={[Function]}
                                tag="p"
                                type="checkbox"
                                value="C"
                            />
                             Create
                        </Label>
                    </FormGroup>
                    <FormGroup
                        check={true}
                        inline={true}
                        tag="div"
                    >
                        <Label
                            check={true}
                            tag="label"
                            widths={
                                Array [
                                    "xs",
                                    "sm",
                                    "md",
                                    "lg",
                                    "xl",
                                  ]
                            }
                        >
                            <Input
                                checked={true}
                                className="operations"
                                id="readOperations"
                                name="operations"
                                onChange={[Function]}
                                tag="p"
                                type="checkbox"
                                value="R"
                            />
                             Read
                        </Label>
                    </FormGroup>
                    <FormGroup
                        check={true}
                        inline={true}
                        tag="div"
                    >
                        <Label
                            check={true}
                            tag="label"
                            widths={
                                Array [
                                    "xs",
                                    "sm",
                                    "md",
                                    "lg",
                                    "xl",
                                  ]
                            }
                        >
                            <Input
                                checked={false}
                                className="operations"
                                id="updateOperations"
                                name="operations"
                                onChange={[Function]}
                                tag="p"
                                type="checkbox"
                                value="U"
                            />
                             Update
                        </Label>
                    </FormGroup>
                    <FormGroup
                        check={true}
                        inline={true}
                        tag="div"
                    >
                        <Label
                            check={true}
                            tag="label"
                            widths={
                                Array [
                                    "xs",
                                    "sm",
                                    "md",
                                    "lg",
                                    "xl",
                                  ]
                            }
                        >
                            <Input
                                checked={false}
                                className="operations"
                                id="deleteOperations"
                                name="operations"
                                onChange={[Function]}
                                tag="p"
                                type="checkbox"
                                value="D"
                            />
                             Delete
                        </Label>
                    </FormGroup>
                </div>
                <div
                    data-field-span="1"
                >
                    <Label
                        tag="label"
                        widths={
                            Array [
                                "xs",
                                "sm",
                                "md",
                                "lg",
                                "xl",
                              ]
                        }
                    >
                        Resource
                    </Label>
                    <Select
                        arrowRenderer={[Function]}
                        autosize={true}
                        backspaceRemoves={true}
                        backspaceToRemoveMessage="Press backspace to remove {label}"
                        clearAllText="Clear all"
                        clearRenderer={[Function]}
                        clearValueText="Clear value"
                        clearable={true}
                        closeOnSelect={true}
                        deleteRemoves={true}
                        delimiter=","
                        disabled={false}
                        escapeClearsValue={true}
                        filterOptions={[Function]}
                        id="resource"
                        ignoreAccents={true}
                        ignoreCase={true}
                        inputProps={Object {}}
                        isLoading={false}
                        joinValues={false}
                        labelKey="label"
                        matchPos="any"
                        matchProp="any"
                        menuBuffer={0}
                        menuRenderer={[Function]}
                        multi={false}
                        name="form-field-name"
                        noResultsText="No results found"
                        onBlurResetsInput={true}
                        onChange={[Function]}
                        onCloseResetsInput={true}
                        onSelectResetsInput={true}
                        openOnClick={true}
                        optionComponent={[Function]}
                        options={
                            Array [
                                Object {
                                  "isAdmin": false,
                                  "label": "ACCOUNT",
                                  "name": "accounts",
                                },
                                Object {
                                  "isAdmin": false,
                                  "label": "ACCOUNT_SUMMARY",
                                  "name": "account-summary",
                                },
                                Object {
                                  "isAdmin": true,
                                  "label": "USER",
                                  "name": "users",
                                },
                                Object {
                                  "isAdmin": true,
                                  "label": "FLAT",
                                  "name": "flats",
                                },
                                Object {
                                  "isAdmin": true,
                                  "label": "RESIDENT",
                                  "name": "residents",
                                },
                                Object {
                                  "isAdmin": true,
                                  "label": "ROLE",
                                  "name": "roles",
                                },
                                Object {
                                  "isAdmin": true,
                                  "label": "PERMISSION",
                                  "name": "permissions",
                                },
                                Object {
                                  "isAdmin": true,
                                  "label": "FLAT_RESIDENT",
                                  "name": "flats-residents",
                                },
                                Object {
                                  "isAdmin": true,
                                  "label": "USER_ROLE",
                                  "name": "users-roles",
                                },
                                Object {
                                  "isAdmin": true,
                                  "label": "ROLE_PERMISSION",
                                  "name": "roles-permissions",
                                },
                                Object {
                                  "isAdmin": false,
                                  "label": "USER_PROFILE",
                                  "name": "user-profile",
                                },
                                Object {
                                  "isAdmin": false,
                                  "label": "BALANCE",
                                  "name": "balance",
                                },
                              ]
                        }
                        pageSize={5}
                        placeholder="Select a Resource..."
                        removeSelected={true}
                        required={false}
                        rtl={false}
                        scrollMenuIntoView={true}
                        searchable={true}
                        simpleValue={true}
                        tabSelectsValue={true}
                        trimFilter={true}
                        value="ACCOUNT"
                        valueComponent={[Function]}
                        valueKey="name"
                    />
                </div>
            </div>
            <div
                data-row-span="1"
            >
                <div
                    data-field-span="1"
                >
                    <Label
                        tag="label"
                        widths={
                            Array [
                                "xs",
                                "sm",
                                "md",
                                "lg",
                                "xl",
                              ]
                        }
                    >
                        Conditions (optional)
                    </Label>
                    <Input
                        className="inputField"
                        id="condition"
                        name="condition"
                        onChange={[Function]}
                        placeholder="<conditional script here>"
                        tag="p"
                        title="Short condition on Permission"
                        type="text"
                        value=""
                    />
                </div>
            </div>
            <div
                data-row-span="1"
            >
                <div
                    data-field-span="1"
                >
                    <Label
                        tag="label"
                        widths={
                            Array [
                                "xs",
                                "sm",
                                "md",
                                "lg",
                                "xl",
                              ]
                        }
                    >
                        Description
                    </Label>
                    <Input
                        className="inputField"
                        id="description"
                        name="description"
                        onChange={[Function]}
                        placeholder="Description here"
                        tag="p"
                        type="text"
                        value="Grands Read-only permission on ACCOUNT"
                    />
                </div>
            </div>
        </fieldset>
        <br />
        <Button
            color="primary"
            disabled={true}
            hidden={false}
            tag="button"
            title="No changes to save"
            type="submit"
        >
            Save
        </Button>
        <Button
            color="link"
            tag="button"
        >
            <Link
                className="text-danger"
                replace={false}
                title="Go to Permissions"
                to="/permissions"
            >
                Cancel
            </Link>
        </Button>
    </Form>
</div>,
]
`;
